commit 34845636a184f3be91a531098192592cbe6db587
From: Andreas Schwab <schwab@linux-m68k.org>
Date: Wed Dec 28 15:57:15 2011 -0800
Subject: [PATCH 7/7] procfs: do not confuse jiffies with cputime64_t
Patch-mainline: 3.2
Reference: BF#152721

    procfs: do not confuse jiffies with cputime64_t
    
    Commit 2a95ea6c0d129b4 ("procfs: do not overflow get_{idle,iowait}_time
    for nohz") did not take into account that one some architectures jiffies
    and cputime use different units.
    
    This causes get_idle_time() to return numbers in the wrong units, making
    the idle time fields in /proc/stat wrong.
    
    Instead of converting the usec value returned by
    get_cpu_{idle,iowait}_time_us to units of jiffies, use the new function
    usecs_to_cputime64 to convert it to the correct unit of cputime64_t.
    
Signed-off-by: Andreas Schwab <schwab@linux-m68k.org>
Acked-by: Michal Hocko <mhocko@suse.cz>
Cc: Arnd Bergmann <arnd@arndb.de>
Cc: "Artem S. Tashkinov" <t.artem@mailcity.com>
Cc: Dave Jones <davej@redhat.com>
Cc: Alexey Dobriyan <adobriyan@gmail.com>
Cc: Thomas Gleixner <tglx@linutronix.de>
Cc: "Luck, Tony" <tony.luck@intel.com>
Cc: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Cc: Martin Schwidefsky <schwidefsky@de.ibm.com>
Cc: Heiko Carstens <heiko.carstens@de.ibm.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Coly Li <bosong.ly@taobao.com>
---
Index: linux-2.6.32-220.7.1.el5/arch/ia64/include/asm/cputime.h
===================================================================
--- linux-2.6.32-220.7.1.el5.orig/arch/ia64/include/asm/cputime.h	2012-03-20 14:40:05.873391483 +0800
+++ linux-2.6.32-220.7.1.el5/arch/ia64/include/asm/cputime.h	2012-03-20 14:40:33.221527050 +0800
@@ -66,6 +66,7 @@
  */
 #define cputime_to_usecs(__ct)		((__ct) / NSEC_PER_USEC)
 #define usecs_to_cputime(__usecs)	((__usecs) * NSEC_PER_USEC)
+#define usecs_to_cputime64(__usecs)	usecs_to_cputime(__usecs)
 
 /*
  * Convert cputime <-> seconds
Index: linux-2.6.32-220.7.1.el5/arch/powerpc/include/asm/cputime.h
===================================================================
--- linux-2.6.32-220.7.1.el5.orig/arch/powerpc/include/asm/cputime.h	2012-03-20 14:40:05.845391362 +0800
+++ linux-2.6.32-220.7.1.el5/arch/powerpc/include/asm/cputime.h	2012-03-20 14:40:33.221527050 +0800
@@ -178,6 +178,8 @@
 	return ct;
 }
 
+#define usecs_to_cputime64(us)		usecs_to_cputime(us)
+
 /*
  * Convert cputime <-> seconds
  */
Index: linux-2.6.32-220.7.1.el5/arch/s390/include/asm/cputime.h
===================================================================
--- linux-2.6.32-220.7.1.el5.orig/arch/s390/include/asm/cputime.h	2012-03-20 14:40:05.857391400 +0800
+++ linux-2.6.32-220.7.1.el5/arch/s390/include/asm/cputime.h	2012-03-20 14:40:33.221527050 +0800
@@ -102,6 +102,8 @@
 	return (cputime_t) m * 4096;
 }
 
+#define usecs_to_cputime64(m)		usecs_to_cputime(m)
+
 /*
  * Convert cputime to milliseconds and back.
  */
Index: linux-2.6.32-220.7.1.el5/fs/proc/stat.c
===================================================================
--- linux-2.6.32-220.7.1.el5.orig/fs/proc/stat.c	2012-03-20 14:40:28.949505881 +0800
+++ linux-2.6.32-220.7.1.el5/fs/proc/stat.c	2012-03-20 14:40:33.221527050 +0800
@@ -33,7 +33,7 @@
 		idle = kstat_cpu(cpu).cpustat.idle;
 		idle = cputime64_add(idle, arch_idle_time(cpu));
 	} else
-		idle = nsecs_to_jiffies64(1000 * idle_time);
+		idle = usecs_to_cputime64(idle_time);
 
 	return idle;
 }
@@ -47,7 +47,7 @@
 		/* !NO_HZ so we can rely on cpustat.iowait */
 		iowait = kstat_cpu(cpu).cpustat.iowait;
 	else
-		iowait = nsecs_to_jiffies64(1000 * iowait_time);
+		iowait = usecs_to_cputime64(iowait_time);
 
 	return iowait;
 }
Index: linux-2.6.32-220.7.1.el5/include/asm-generic/cputime.h
===================================================================
--- linux-2.6.32-220.7.1.el5.orig/include/asm-generic/cputime.h	2012-03-20 14:40:21.549469222 +0800
+++ linux-2.6.32-220.7.1.el5/include/asm-generic/cputime.h	2012-03-20 14:40:33.221527050 +0800
@@ -52,6 +52,7 @@
  */
 #define cputime_to_usecs(__ct)		jiffies_to_usecs(__ct)
 #define usecs_to_cputime(__msecs)	usecs_to_jiffies(__msecs)
+#define usecs_to_cputime64(__msecs)	nsecs_to_jiffies64((__msecs) * 1000)
 
 /*
  * Convert cputime to seconds and back.
